### DISCUSSION TITLE
请问nginx反代grpc。xray获取访问者真实IP，如何修改配置文件？

### DISCUSSION THREAD

**Discussion by ghost** [XTLS/Xray-core#462] [Q&A] [answered]
xray 版本1.4.1

nginx 配置
`    server {
        listen 443 ssl http2;

        ssl_certificate_key /etc/ssl/private/key.pem;
        ssl_certificate /etc/ssl/private/fullchain.pem;
		
        ssl_protocols TLSv1.2 TLSv1.3;
        ssl_ciphers TLS13_AES_128_GCM_SHA256:TLS13_AES_256_GCM_SHA384:TLS13_CHACHA20_POLY1305_SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305;
        ssl_prefer_server_ciphers on;

        ssl_session_cache shared:SSL:1m;
        ssl_session_timeout 5m;
        ssl_session_tickets on;

        add_header Strict-Transport-Security "max-age=31536000; includeSubDomains" always;

        location /lovelive/ {
            if ($request_method != "POST") {
               return 404;
            }
            client_max_body_size 0;
            client_body_timeout 1h;
            grpc_read_timeout 1h;

            grpc_pass grpc://127.0.0.1:8443;
        }

        location / {
            proxy_pass https://www.lovelive-anime.jp/;
        }
    }

xray 服务器端配置

{
    "log": {
        "loglevel": "warning",
        "access": "/var/log/xray/access.log",
        "error": "/var/log/xray/error.log"
    },
    "inbounds": [
        {
            "listen": "127.0.0.1",
            "port": 8443,
            "protocol": "vless",
            "settings": {
                "clients": [
                    {
                        "id": "chika"
                    }
                ],
                "decryption": "none"
            },
            "streamSettings": {
                "network": "grpc",
                "grpcSettings": {
                    "serviceName": "lovelive"
                }
            }
        }
    ],
    "outbounds": [
        {
            "protocol": "freedom"
        }
    ]
}
`

小白试过在nginx里加
grpc_set_header X-Real-IP $http_x_forwarded_for;
grpc_set_header X-Forwarded-For $proxy_protocol_addr;
			

xray日志里还是显示0.0.0.0

**Comment by LearZhou** [comment]
用HaProxy反代经过cf的ws，可以直接读取cf header中的CF-Connecting-IP，获得客户端真实ip；但是，过cf的gRPC不知道怎么获取这个ip，似乎跟http/1.1不太一样，直接套用不好使。

**Reply by ghost** [reply]
我不过CDN，直连到VPS。我只是想在xray的日志里，记录下访问者的IP地址。我在TG群里看别人问也没得到有用的方法。小白就来问问了。

**Comment by ghost** [comment, marked as answer]
可以使用 Proxy Protocol 传递客户端 IP 地址。

注：
- 仅在 v1.4.2 后，gRPC 才会获取远程地址。
- acceptProxyProtocol 需要写在 sockopt 中。

**Reply by ghost** [reply]
感谢回复

nginx配置是

        location /lovelive/ {
            if ($request_method != "POST") {
               return 404;
            }
            client_body_timeout 1h;
            grpc_read_timeout 1h;
            grpc_pass grpc://127.0.0.1:8443;
            grpc_set_header X-Real-IP $http_x_forwarded_for;
            grpc_set_header X-Forwarded-For $proxy_protocol_addr;
        }


xray配置是

    "inbounds": [
        {
            "listen": "127.0.0.1",
            "port": 8443,
            "protocol": "vless",
            "settings": {
                "clients": [
                    {
                        "id": "chika"
                    }
                ],
                "decryption": "none"
            },
            "streamSettings": {
                "network": "grpc",
                "grpcSettings": {
                    "serviceName": "lovelive"
                },
                "sockopt": {
                    "acceptProxyProtocol": true
                }
            }
        }
    ],



连接失败。请问参数哪有问题？

**Reply by ghost** [reply]
似乎 nginx 配置里并没有使用 Proxy Protocol。

但我不是很确定 ，可以问一下其他人

**Reply by ghost** [reply]
谢谢你。大佬晚上辛苦了。

**Reply by 0oook** [reply]
在nginx里加
grpc_set_header X-Real-IP $remote_addr; 
grpc_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
